### PR TITLE
Gaussian nll loss scalar variance support

### LINKED ISSUE #138747: Speed up gaussian_nll_loss when variance is always the same scalar value
### 🚀 The feature, motivation and pitch

I've been using `gaussian_nll_loss` and noticed that it can be quite slow, due to the check for non-negative variances here: https://github.com/pytorch/pytorch/blob/main/torch/nn/functional.py#L3295, which forces a gpu-cpu sync. See the following screenshot from a profile:
gaussian_nll_loss
This was reliably costing ~10ms for me (admittedly, that includes the some overhead of the profiler).

In my case, the variance I was using was essentially `0.05*torch.ones_like(inputs)` and since it was known to be > 0 ahead of time, the slow check was redundant.

It would be nice if `gaussian_nll_loss` accepted either a Tensor or a float for `var`. If it's a float, have a check on the value and then convert it to a Tensor to avoid the expensive sync, otherwise perform the check as normal. eg at the beginning of the function:
```python
if isinstance(var, float):
   if var < 0:
        raise ValueError("var has negative entry/entries")
    var = var*torch.ones_like(inputs)
elif torch.any(var < 0):
    raise ValueError("var has negative entry/entries")
```

### Alternatives

Alternatively, I was wondering if the non-negative check could be removed, as it is documented that `variance` should be positive.
The concern is that it could be silently set to the (positive) value of `eps` due to the `clamp` call here: https://github.com/pytorch/pytorch/blob/main/torch/nn/functional.py#L3301

### Additional context

If you are happy with the proposed change, I'd be happy to raise a PR for it.

### PR DESCRIPTION
**Pull request by michael-diggin**
Fixes #138747

Adds support for `variance` being a Tensor or a float in `gaussian_nll_loss` to avoid a cpu-gpu sync point in the loss function, when the variance is a static tensor like `<scalar>*torch.ones_like(input)`

### CODE REVIEW

**Review by mikaylagawarecki — APPROVED** [review]
Thanks

**Comment by mikaylagawarecki** [comment]
Seems like CI is failing

**Comment by michael-diggin** [comment]
> Seems like CI is failing

I think it wasn't liking the fact I had used `Tensor | float` as that isn't supported on Python 3.9. I've changed it to use `Union[Tensor, Float]` which should be the exact same.

**Comment by michael-diggin** [comment]
@mikaylagawarecki I've updated the PR which should fix the failing CI, if you would be able to re-trigger that workflow if you're happy with the update?

**Comment by michael-diggin** [comment]
Thanks! I think the only remaining failure is the bc-lint here: https://github.com/pytorch/pytorch/actions/runs/11784651031/job/32831020347?pr=138931#step:2:833
Which I believe is a false positive, as `Tensor -> Union[Tensor, Float]` is a backwards compatible change.
If you agree we can add the `suppress-bc-linter` label and I think that should raise them as notifications rather than errors?

**Comment by michael-diggin** [comment]
> Thanks! I think the only remaining failure is the bc-lint here: https://github.com/pytorch/pytorch/actions/runs/11784651031/job/32831020347?pr=138931#step:2:833 Which I believe is a false positive, as `Tensor -> Union[Tensor, Float]` is a backwards compatible change. If you agree we can add the `suppress-bc-linter` label and I think that should raise them as notifications rather than errors?

Hi @mikaylagawarecki, would you be able to take a look at this comment if you get a chance? I believe adding that label (following the docs [here)](https://github.com/pytorch/test-infra/wiki/BC-Linter#suppression) and rerunning the bc-lint check should fix the last remaining failing check. Thanks! 🙏
